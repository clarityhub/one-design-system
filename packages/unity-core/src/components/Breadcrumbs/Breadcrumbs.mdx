---
menu: Components
name: Breadcrumbs
---

import { Playground, PropsTable } from 'docz';
import Breadcrumbs, { BreadcrumbsPropTypes } from './Breadcrumbs';

# Breadcrumbs

## \<Breadcrumbs /\> component

```js
import Breadcrumbs from '@clarityhub/unity-core/components/Breadcrumbs';
```

## Prop Types

<PropsTable of={BreadcrumbsPropTypes} />

## Examples

<Playground>
    <Breadcrumbs
        crumbs={[
            {
                title: 'Home'
            }
        ]}
    />
    <Breadcrumbs
        crumbs={[
            {
                title: 'Home',
                path: '/',
            },
            {
                title: 'Admin',
            },
        ]}
    />
    <Breadcrumbs
        crumbs={[
            {
                title: 'Home',
                path: '/',
            },
            {
                title: 'Admin',
                path: '/admin',
            },
            {
                title: 'Users',
                // no path
            }
        ]}
    />
</Playground>