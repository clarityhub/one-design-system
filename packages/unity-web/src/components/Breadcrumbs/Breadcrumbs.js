import React from 'react';
import styled from '@emotion/styled';
import { arrayOf, shape, func, string } from 'prop-types';
import { variants } from '@clarityhub/unity-core/lib/typography';
import colors from '@clarityhub/unity-core/lib/colors';

import Link from '../Link';

const BreadcrumbsWrapper = styled.nav`
    ${variants.text.string}

    color: ${colors.black.default};
`;

const BreadcrumbList = styled.ol`
    display: flex;
    flex-wrap: wrap;
    padding: .75rem 1rem;
    margin-bottom: 1rem;
    list-style: none;
`;

const Breadcrumb = styled.li`
    & + &:before {
        display: inline-block;
        padding-right: .5rem;
        padding-left: .5rem;
        color: ${colors.gray.default};
        content: "/";
    }

    a,
    a:visited {
        color: ${colors.black.default};
    }
`;

const defaultLinkRenderer = ({ path, children }) => (
	<Link href={path}>{children}</Link>
);

const Breadcrumbs = ({ crumbs, linkRenderer = defaultLinkRenderer }) => {
	return (
		<BreadcrumbsWrapper ariaLabel="breadcrumb">
			<BreadcrumbList>
				{crumbs.map((crumb, i) => {
					const isCurrent = i === (crumbs.length - 1);
            
					return (
						<Breadcrumb key={i} ariaCurrent={isCurrent ? 'page' : false}>
							{crumb.path ? linkRenderer({
								path: crumb.path,
								children: crumb.title,
							}) : crumb.title}
						</Breadcrumb>
					);
				})}
			</BreadcrumbList>
		</BreadcrumbsWrapper>
	);
};

/**
 * Hack for docz and react-docgen
 */
const BreadcrumbsPropTypes = () => <div />;
BreadcrumbsPropTypes.propTypes = {
	crumbs: arrayOf(shape({
		title: string.isRequired,
		path: string,
	})),
	linkRenderer: func,
};
Breadcrumbs.propTypes = BreadcrumbsPropTypes.propTypes;
export { BreadcrumbsPropTypes };

export default Breadcrumbs;
