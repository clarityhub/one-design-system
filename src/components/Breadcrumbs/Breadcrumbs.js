import React from 'react';
import styled from '@emotion/styled';

import Link from '../Link';
import { variants } from '../../theme/fonts';
import colors from '../../theme/colors';
import { arrayOf, shape, func, string } from 'prop-types';

const BreadcrumbsWrapper = styled.nav`
    ${variants.text}

    color: ${colors.black};
`;

const BreadcrumbList = styled.ol`
    display: flex;
    flex-wrap: wrap;
    padding: .75rem 1rem;
    margin-bottom: 1rem;
    list-style: none;
`;

const Breadcrumb = styled.li`
    & + &:before {
        display: inline-block;
        padding-right: .5rem;
        padding-left: .5rem;
        color: ${colors.gray};
        content: "/";
    }

    a,
    a:visited {
        color: ${colors.black};
    }
`;

const defaultLinkRenderer = ({ path, children }) => (
	<Link href={path}>{children}</Link>
);

const Breadcrumbs = ({ crumbs, linkRenderer = defaultLinkRenderer }) => {
	return (
		<BreadcrumbsWrapper ariaLabel="breadcrumb">
			<BreadcrumbList>
				{crumbs.map((crumb, i) => {
					const isCurrent = i === (crumbs.length - 1);
            
					return (
						<Breadcrumb key={i} ariaCurrent={isCurrent ? 'page' : false}>
							{crumb.path ? linkRenderer({
								path: crumb.path,
								children: crumb.title,
							}) : crumb.title}
						</Breadcrumb>
					);
				})}
			</BreadcrumbList>
		</BreadcrumbsWrapper>
	);
};

/**
 * Hack for docz and react-docgen
 */
const BreadcrumbsPropTypes = () => <div />;
BreadcrumbsPropTypes.propTypes = {
	crumbs: arrayOf(shape({
		title: string.isRequired,
		path: string,
	})),
	linkRenderer: func,
};
Breadcrumbs.propTypes = BreadcrumbsPropTypes.propTypes;
export { BreadcrumbsPropTypes };

export default Breadcrumbs;
