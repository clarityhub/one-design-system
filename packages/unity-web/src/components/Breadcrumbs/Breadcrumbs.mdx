---
menu: Components
name: Breadcrumbs
---

import { Playground, PropsTable } from 'docz';
import Breadcrumbs, { BreadcrumbsPropTypes } from './Breadcrumbs';

# Breadcrumbs

Use breadcrumbs to show a user where they are in the sitemap.

```js
import Breadcrumbs from '@clarityhub/unity-web/lib/components/Breadcrumbs';
```

## Example

<Playground>
    <Breadcrumbs
        crumbs={[
            {
                title: 'Home'
            }
        ]}
    />
    <Breadcrumbs
        crumbs={[
            {
                title: 'Home',
                path: '/',
            },
            {
                title: 'Admin',
            },
        ]}
    />
    <Breadcrumbs
        crumbs={[
            {
                title: 'Home',
                path: '/',
            },
            {
                title: 'Admin',
                path: '/admin',
            },
            {
                title: 'Users',
                // no path
            }
        ]}
    />
</Playground>

## Link Renderer

By default, the breadcrumbs will use a Link to render the given path.

You can override how breadcrumbs are rendered (so you can use React Router Links for example) by providing a `linkRenderer`.

The default renderer looks like:

```js
const defaultLinkRenderer = ({ path, children }) => (
	<Link href={path}>{children}</Link>
);
```

## Prop Types

The following prop types are available for the Breadcrumbs component.

<PropsTable of={BreadcrumbsPropTypes} />