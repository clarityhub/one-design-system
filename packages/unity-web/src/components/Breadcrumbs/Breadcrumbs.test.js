import React from 'react';
import renderer from 'react-test-renderer';

import Breadcrumbs from './Breadcrumbs';

describe('<Breadcrumbs />', () => {
	it('renders crumbs', () => {
		const node = (
			<div>
				<Breadcrumbs
					crumbs={[
						{
							title: 'Home',
						},
					]}
				/>
				<Breadcrumbs
					crumbs={[
						{
							title: 'Home',
							path: '/',
						},
						{
							title: 'Admin',
						},
					]}
				/>
				<Breadcrumbs
					crumbs={[
						{
							title: 'Home',
							path: '/',
						},
						{
							title: 'Admin',
							path: '/admin',
						},
						{
							title: 'Long interview name – October 3rd, 2019',
							// no path
						},
					]}
				/>
			</div>
		);
		const tree = renderer.create(node).toJSON();

		expect(tree).toMatchSnapshot();
	});
});