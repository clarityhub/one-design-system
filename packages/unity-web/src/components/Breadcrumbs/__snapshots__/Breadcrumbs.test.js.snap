// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Breadcrumbs /> renders crumbs 1`] = `
.emotion-4 {
  font-family: 'Open Sans',Arial,sans-serif;
  font-weight: 400;
  color: #333;
  max-width: 100%;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: .75rem 1rem;
  list-style: none;
  white-space: nowrap;
}

.emotion-0 {
  display: inline-block;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  min-width: 3rem;
}

.emotion-0:first-of-type {
  overflow: inherit;
}

.emotion-0 + .emotion-0:before {
  display: inline-block;
  padding-right: .5rem;
  padding-left: .5rem;
  color: #C2C2C2;
  content: "/";
}

.emotion-0 a,
.emotion-0 a:visited {
  color: #333;
}

.emotion-6 {
  border-bottom: 1px solid #566ad2;
  color: #566ad2;
  display: inline-block;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-family: 'Open Sans',Arial,sans-serif;
  font-weight: 400;
  string: font-family:'Open Sans',Arial,sans-serif;
  font-weight: 400;
  color: #566ad2;
}

.emotion-6:hover,
.emotion-6:active {
  border-bottom: 2px solid #566ad2;
  border-bottom-width: 2px;
  color: #566ad2;
  margin-bottom: -1px;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.emotion-6:visited {
  border-bottom: 1px solid rgb(50,72,187);
  color: rgb(50,72,187);
  -webkit-text-decoration: none;
  text-decoration: none;
}

.emotion-6:visited:hover,
.emotion-6:visited:active {
  margin-bottom: 0;
}

<div>
  <nav
    className="emotion-4 emotion-5"
  >
    <ol
      className="emotion-2 emotion-3"
    >
      <li
        className="emotion-0 emotion-1"
      >
        Home
      </li>
    </ol>
  </nav>
  <nav
    className="emotion-4 emotion-5"
  >
    <ol
      className="emotion-2 emotion-3"
    >
      <li
        className="emotion-0 emotion-1"
      >
        <a
          className="emotion-6"
          href="/"
        >
          Home
        </a>
      </li>
      <li
        className="emotion-0 emotion-1"
      >
        Admin
      </li>
    </ol>
  </nav>
  <nav
    className="emotion-4 emotion-5"
  >
    <ol
      className="emotion-2 emotion-3"
    >
      <li
        className="emotion-0 emotion-1"
      >
        <a
          className="emotion-6"
          href="/"
        >
          Home
        </a>
      </li>
      <li
        className="emotion-0 emotion-1"
      >
        <a
          className="emotion-6"
          href="/admin"
        >
          Admin
        </a>
      </li>
      <li
        className="emotion-0 emotion-1"
      >
        Long interview name – October 3rd, 2019
      </li>
    </ol>
  </nav>
</div>
`;
